argh/tests/lib.rs

```diff
@@ -94,6 +94,369 @@ fn subcommand_example() {
+    struct SubCommandOne {
+        #[argh(option)]
+        /// how many x
+        x: usize,
+    }
+
+    #[derive(FromArgs, PartialEq, Debug)]
+    /// Second subcommand.
+    #[argh(subcommand, name = "two")]
+    struct SubCommandTwo {
+        #[argh(switch)]
+        /// whether to fooey
+        fooey: bool,
+    }
+
+    let one = TopLevel::from_args(&["cmdname"], &["one", "--x", "2"]).expect("sc 1");
+    assert_eq!(one, TopLevel { nested: Some(MySubCommandEnum::One(SubCommandOne { x: 2 })) },);
+}
+
+#[test]
+fn optional_subcommand_empty_show_help_no_default() {
+    #[derive(FromArgs, PartialEq, Debug)]
+    /// Top-level command.
+    struct TopLevel {
+        #[argh(subcommand)]
+        nested: Option<MySubCommandEnum>,
+    }
+
+    #[derive(FromArgs, PartialEq, Debug)]
+    #[argh(subcommand)]
+    enum MySubCommandEnum {
+        One(SubCommandOne),
+        Two(SubCommandTwo),
+    }
+
+    #[derive(FromArgs, PartialEq, Debug)]
+    /// First subcommand.
+    #[argh(subcommand, name = "one")]
+    struct SubCommandOne {
+        #[argh(option)]
+        /// how many x
+        x: usize,
+    }
+
+    #[derive(FromArgs, PartialEq, Debug)]
+    /// Second subcommand.
+    #[argh(subcommand, name = "two")]
+    struct SubCommandTwo {
+        #[argh(switch)]
+        /// whether to fooey
+        fooey: bool,
+    }
+
+    let help_str = r###"Usage: cmdname [<command>] [<args>]
+
+Top-level command.
+
+Options:
+  --help            display usage information
+
+Commands:
+  one               First subcommand.
+  two               Second subcommand.
+"###;
+
+    match TopLevel::from_args(&["cmdname"], &["--help"]) {
+        Ok(_) => panic!("help was parsed as args"),
+        Err(e) => {
+            assert_eq!(help_str, e.output);
+            e.status.expect("help returned an error");
+        }
+    }
+}
+
+#[test]
+fn optional_subcommand_empty_show_help_for_both() {
+    #[derive(FromArgs, PartialEq, Debug)]
+    /// Top-level command.
+    struct TopLevel {
+        #[argh(subcommand, default = "one")]
+        nested: Option<MySubCommandEnum>,
+    }
+
+    #[derive(FromArgs, PartialEq, Debug)]
+    #[argh(subcommand)]
+    enum MySubCommandEnum {
+        One(SubCommandOne),
+        Two(SubCommandTwo),
+    }
+
+    #[derive(FromArgs, PartialEq, Debug)]
+    /// First subcommand.
+    #[argh(subcommand, name = "one")]
+    struct SubCommandOne {
+        #[argh(option)]
+        /// how many x
+        x: usize,
+    }
+
+    #[derive(FromArgs, PartialEq, Debug)]
+    /// Second subcommand.
+    #[argh(subcommand, name = "two")]
+    struct SubCommandTwo {
+        #[argh(switch)]
+        /// whether to fooey
+        fooey: bool,
+    }
+
+    let help_str = r###"Usage: cmdname [<command>] [<args>]
+
+Top-level command.
+
+Options:
+  --help            display usage information
+
+Commands:
+  one               First subcommand.
+  two               Second subcommand.
+
+Usage: cmdname one --x <x>
+
+First subcommand.
```

Comment:
Is it possible to not print these two lines? The goal is to instruct the user of `cmdname` that they can use the commands `one` and `two` but that by default the command `one` would be called and immediately surface the options of that command. I think these two lines make this a bit confusing. Could we instead do something like:

```
Top-level command.

Options:
  --help            display usage information
  
Commands:
  one               First subcommand.
  two               Second subcommand.

When no command is given, the commad `one` is the default. Allowing the additional:

Options:
   --x               how many x
  --help            display usage information
```

Reply:
This is a bit harder to achieve given the state of the code, I can add the line of 

`When no command is given, the commad `one` is the default. Allowing the additional`

And do this in a later patch if that is okay with you

Reply:
I did implement this in the end but it took a lot more code than what I'd like to do it.